### ISSUE TITLE
Change template function definition to support Visual Studio

### ISSUE THREAD

**Issue by MonkeyDied** [robotology/osqp-eigen#22]
When compiling the mpc example using VS2013 on Win10, the following error occurred.
> fatal error C1001: An internal error has occurred in the compiler.
>  (compiler file 'msc1.cpp', line 1325)
> ...

After some testing I found that it seems to be a problem with the template parsing for functions of the following form.
`template<int n> ret_type some_function(Eigen::Matrix<c_float, n, 1>& arg)`

This could be remedied by changing the definition to the following.
`template<class T, int n> ret_type some_function(Eigen::Matrix<T, n, 1>& arg)`

For example, in Data.hpp & Data.tpp, there is a function called setGradient. The current version and the suggested revised version are as follows.
`/*current*/ template<int n> bool Data::setGradient(Eigen::Matrix<c_float, n, 1>& gradient)`
`/*suggest*/ template<class T, int n> bool Data::setGradient(Eigen::Matrix<T, n, 1>& gradient)`

Below are all the functions I found to have the same problems.
For Data.hpp & Data.tpp
`template<int n> bool Data::setGradient(Eigen::Matrix<c_float, n, 1>& gradientVector);`
`template<int m>  bool setLowerBound(Eigen::Matrix<c_float, m, 1>& lowerBoundVector);`
`template<int m>  bool setUpperBound(Eigen::Matrix<c_float, m, 1>& upperBoundVector);`
For Solver.hpp & Solver.tpp
`template<int n>   bool updateGradient(Eigen::Matrix<c_float, n, 1>& gradient);`
`template<int m>  bool updateLowerBound(Eigen::Matrix<c_float, m, 1>& lowerBound);`
`template<int m>  bool updateUpperBound(Eigen::Matrix<c_float, m, 1>& upperBound);`
`template<int m>  bool updateBounds(Eigen::Matrix<c_float, m, 1>& lowerBound,                          Eigen::Matrix<c_float, m, 1>& upperBound);`

It would be nice if someone could help fix all the definitions.
Thx~

**Comment by GiulioRomualdi** [comment]
Hi @MonkeyDied, first of all, thanks for using `osqp-eigen`. 
Regarding your question, I think that the problem is related to the _template argument deduction_. Can you copy the lines of the code that cause the issue?

**Comment by MonkeyDied** [comment]
@GiulioRomualdi The following are the definitions of the functions that cause the issue.

https://github.com/robotology/osqp-eigen/blob/a74f46f90f6f93d0b125909ce54930a66e023412/include/OsqpEigen/Data.hpp#L93-L94

https://github.com/robotology/osqp-eigen/blob/a74f46f90f6f93d0b125909ce54930a66e023412/include/OsqpEigen/Data.hpp#L104-L118

https://github.com/robotology/osqp-eigen/blob/a74f46f90f6f93d0b125909ce54930a66e023412/include/OsqpEigen/Solver.hpp#L125-L157

**Comment by GiulioRomualdi** [comment]
Hi @MonkeyDied, I was referring to your code :smiley:

**Comment by MonkeyDied** [comment]
Hi @GiulioRomualdi , I encounter the problem when compiling the MPCExample using Visual Studio 2013 in Win10. The Eigen version is 3.3.5. osqp version is 0.5.0. 

The following codes produce a compile time fatal error C1001 as described in the first post.
(The setLinearConstraintsMatrix function works fine)

https://github.com/robotology/osqp-eigen/blob/a74f46f90f6f93d0b125909ce54930a66e023412/example/src/MPCExample.cpp#L261-L264

https://github.com/robotology/osqp-eigen/blob/a74f46f90f6f93d0b125909ce54930a66e023412/example/src/MPCExample.cpp#L293

**Comment by GiulioRomualdi** [comment]
Can you try to edit the lines in the following way?

```cpp
 if(!solver.data()->setGradient<Eigen::Dynamic>(gradient)) return 1; 
 if(!solver.data()->setLinearConstraintsMatrix(linearMatrix)) return 1; 
 if(!solver.data()->setLowerBound<Eigen::Dynamic>(lowerBound)) return 1; 
 if(!solver.data()->setUpperBound<Eigen::Dynamic>(upperBound)) return 1; 
```

and 
```cpp
 if(!solver.updateBounds<Eigen::Dynamic>(lowerBound, upperBound)) return 1; 
```

**Comment by MonkeyDied** [comment]
Hi @GiulioRomualdi , that seems to solve the problem.

**Comment by GiulioRomualdi** [comment]
Nice! I'll fix it!

**Comment by GiulioRomualdi** [comment]
Hi @MonkeyDied this PR (#24 ) should solve your problem. Let me know if you have still problems.

**Comment by GiulioRomualdi** [comment]
Fixed in #24
